lib/dao.js

```diff
@@ -332,14 +332,64 @@ DataAccessObject.findOrCreate = function findOrCreate(query, data, callback) {
     };
   }
 
-  var t = this;
-  this.findOne(query, function (err, record) {
-    if (err) return callback(err);
-    if (record) return callback(null, record, false);
-    t.create(data, function (err, record) {
-      callback(err, record, record != null);
+  var self = this;
+
+  if (this.getDataSource().connector.findOrCreate) {
+
+    try {
+      this._normalize(query);
+    } catch (err) {
+      return process.nextTick(function () {
+        callback && callback(err);
+      });
+    }
+
+    this.applyScope(query);
+
+    var enforced = {};
+    var Model = this.lookupModel(data);
+    var obj = data instanceof Model ? data : new Model(data);
+
+    this.applyProperties(enforced, obj);
+    obj.setAttributes(enforced);
+
+    data = obj.toObject(true);
+
+    // validation required
+    obj.isValid(function (valid) {
+      if (valid) {
+        findOrCreate();
+      } else {
+        callback(new ValidationError(obj), obj);
+      }
+    }, data);
```

Comment:
This looks like a lot of copy-and-pasted code, is there a way how to refactor it into a shared method?